zk-keygen/src/main.rs

```diff
@@ -0,0 +1,371 @@
+        match self {
+            KeyType::ElGamal => "elgamal.json".to_string(),
+            KeyType::Symmetric => "symmetric.json".to_string(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        solana_sdk::pubkey::Pubkey,
+        tempfile::{tempdir, TempDir},
+    };
+
+    fn process_test_command(args: &[&str]) -> Result<(), Box<dyn error::Error>> {
+        let solana_version = solana_version::version!();
+        let app_matches = app(solana_version).get_matches_from(args);
+        do_main(&app_matches)
+    }
+
+    fn tmp_outfile_path(out_dir: &TempDir, name: &str) -> String {
+        let path = out_dir.path().join(name);
+        path.into_os_string().into_string().unwrap()
+    }
+
+    #[test]
+    fn test_new_elgamal() {
+        let outfile_dir = tempdir().unwrap();
+        // use `Pubkey::new_unique()` to generate names for temporary key files
+        let outfile_path = tmp_outfile_path(&outfile_dir, &Pubkey::new_unique().to_string());
+
+        // general success case
+        process_test_command(&[
+            "solana-zk-keygen",
+            "new",
+            "--key-type",
+            "elgamal",
+            "--outfile",
+            &outfile_path,
+            "--no-bip39-passphrase",
+        ])
+        .unwrap();
+
+        // refuse to overwrite file
+        let result = process_test_command(&[
+            "solana-zk-keygen",
+            "new",
+            "--key-type",
+            "elgamal",
+            "--outfile",
+            &outfile_path,
+            "--no-bip39-passphrase",
+        ])
+        .unwrap_err()
+        .to_string();
+
+        let expected = format!("Refusing to overwrite {outfile_path} without --force flag");
+        assert_eq!(result, expected);
+
+        // no outfile
+        process_test_command(&[
+            "solana-keygen",
+            "new",
+            "--key-type",
+            "elgamal",
+            "--no-bip39-passphrase",
+            "--no-outfile",
+        ])
+        .unwrap();
+
+        // sanity check on languages and word count combinations
+        let languages = [
+            "english",
+            "chinese-simplified",
+            "chinese-traditional",
+            "japanese",
+            "spanish",
+            "korean",
+            "french",
+            "italian",
+        ];
+        let word_counts = ["12", "15", "18", "21", "24"];
+
+        for language in languages {
+            for word_count in word_counts {
+                process_test_command(&[
+                    "solana-keygen",
+                    "new",
+                    "--key-type",
+                    "elgamal",
+                    "--no-outfile",
+                    "--no-bip39-passphrase",
+                    "--language",
+                    language,
+                    "--word-count",
+                    word_count,
+                ])
+                .unwrap();
+            }
+        }
```

Comment:
seems like a test for these args would fit better where the trait is declared than being repeated everywhere it's used

Reply:
Yeah I agree. It would be much better to add tests in the clap-v3-utils itself, but it seems tricky to add tests for incomplete cli tools. I can think more about the best way to test these, but it would probably belong in a different PR.

For this specific PR, do you think it is worth keeping these checks? Things like https://github.com/solana-labs/solana/pull/24686 does happen, so I guess it could act as a sanity check...

Reply:
dropping them here is probably fine. they're testing `clap` more than they are our tooling anyway.

as for the testing in the trait context.  it should be pretty trivial to just toss together a featureless mock `clap::App` and call the trait function on it to add the args under test.

Reply:
Okay, makes sense. I removed the checks from the tests!